Use group.WORLD appropriately in process group initialization.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48767 Use group.WORLD appropriately in process group initialization.**

As part of investigating
https://github.com/pytorch/pytorch/issues/48464, I realized some weird
inconsistency in how we use `_default_pg` and `group.WORLD`. `group.WORLD`
apparently was an `object()` and never changed despite `_default_pg` changing.
In this sense, `group.WORLD` was being used a constant to refer to the default
pg, but wasn't of type PG at all. In fact the passed in group is also compared
via `==` to `group.WORLD` in many places, and it just worked since the default
argument was `group.WORLD`.

To clean this up, I got rid of `_default_pg` completely and instead used
`group.WORLD` as the default pg throughout the codebase. This also fixes the
documentation issues mentioned in
https://github.com/pytorch/pytorch/issues/48464.

#Closes: https://github.com/pytorch/pytorch/issues/48464

Differential Revision: [D25292893](https://our.internmc.facebook.com/intern/diff/D25292893/)

pritamdamania committed:
Use group.WORLD appropriately in process group initialization.

As part of investigating
https://github.com/pytorch/pytorch/issues/48464, I realized some weird
inconsistency in how we use `_default_pg` and `group.WORLD`. `group.WORLD`
apparently was an `object()` and never changed despite `_default_pg` changing.
In this sense, `group.WORLD` was being used a constant to refer to the default
pg, but wasn't of type PG at all. In fact the passed in group is also compared
via `==` to `group.WORLD` in many places, and it just worked since the default
argument was `group.WORLD`.

To clean this up, I got rid of `_default_pg` completely and instead used
`group.WORLD` as the default pg throughout the codebase. This also fixes the
documentation issues mentioned in
https://github.com/pytorch/pytorch/issues/48464.

#Closes: https://github.com/pytorch/pytorch/issues/48464

Differential Revision: [D25292893](https://our.internmc.facebook.com/intern/diff/D25292893/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -147,7 +147,7 @@ def __getattribute__(self, key):
 
 class group(object):
-    WORLD = object()
+    WORLD: Optional[ProcessGroup] = None
```
Does this make the below `GroupMember` class sort of strange in that it is like an enum, but GroupMember.WORLD is a ProcessGroup and GroupMember.NON_GROUP_MEMBER is an object()?

rohan-varma replied:
Nit: maybe add a comment saying that it will be the default pg once initialized?

pritamdamania87 replied:
> Does this make the below GroupMember class sort of strange in that it is like an enum, but GroupMember.WORLD is a ProcessGroup and GroupMember.NON_GROUP_MEMBER is an object()?

Yes, it does although this is unfortunately due to how the code was initially written and its hard to change some of this due to backward compatibility reasons. Both classes `group` and `GroupMember` are public and are used by PyTorch users. `NON_GROUP_MEMBER` is particularly weird for the fact that we actually return that to `new_group` calls which are not part of the new_group. As a result, those calls don't get ProcessGroup returned (or None) but this arbitrarty object `NON_GROUP_MEMBER` that is used elsewhere to perform checks.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -316,7 +315,7 @@ def _get_default_group():
         raise RuntimeError("Default process group has not been initialized, "
                            "please make sure to call init_process_group.")
-    return _default_pg
+    return group.WORLD
```
Nit (was also true before this PR): It sort of seems like this function and `_check_default_pg` sort of do the same thing, return the default pg or throw if it is None. Should we dedupe them?

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -816,7 +819,8 @@ class P2POp(object):
         tensor (Tensor): Tensor to send or receive.
         peer (int): Destination or source rank.
-        group (ProcessGroup, optional): The process group to work on.
+        group (ProcessGroup, optional): The process group to work on. If None,
```
Hmm, not sure if I understand how this worked previously? If `group` passed in was always a process group, how could it be compared successfully against `group.WORLD` when it was an `object()` before this PR?

pritamdamania87 replied:
I was surprised by this too and it is actually broken in some cases. In the latest update to the PR I've added a test called `test_pass_default_pg`, which passes with this PR but fails on master. I guess technically you can't really retrieve the default process group unless you use the private `_get_default_group` API, so I guess we were relying on that for this not to break. But I feel that is very fragile.

pritamdamania committed:
Update on "Use group.WORLD appropriately in process group initialization."


As part of investigating
https://github.com/pytorch/pytorch/issues/48464, I realized some weird
inconsistency in how we use `_default_pg` and `group.WORLD`. `group.WORLD`
apparently was an `object()` and never changed despite `_default_pg` changing.
In this sense, `group.WORLD` was being used a constant to refer to the default
pg, but wasn't of type PG at all. In fact the passed in group is also compared
via `==` to `group.WORLD` in many places, and it just worked since the default
argument was `group.WORLD`.

To clean this up, I got rid of `_default_pg` completely and instead used
`group.WORLD` as the default pg throughout the codebase. This also fixes the
documentation issues mentioned in
https://github.com/pytorch/pytorch/issues/48464.

#Closes: https://github.com/pytorch/pytorch/issues/48464

Differential Revision: [D25292893](https://our.internmc.facebook.com/intern/diff/D25292893/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use group.WORLD appropriately in process group initialization."


As part of investigating
https://github.com/pytorch/pytorch/issues/48464, I realized some weird
inconsistency in how we use `_default_pg` and `group.WORLD`. `group.WORLD`
apparently was an `object()` and never changed despite `_default_pg` changing.
In this sense, `group.WORLD` was being used a constant to refer to the default
pg, but wasn't of type PG at all. In fact the passed in group is also compared
via `==` to `group.WORLD` in many places, and it just worked since the default
argument was `group.WORLD`.

To clean this up, I got rid of `_default_pg` completely and instead used
`group.WORLD` as the default pg throughout the codebase. This also fixes the
documentation issues mentioned in
https://github.com/pytorch/pytorch/issues/48464.

#Closes: https://github.com/pytorch/pytorch/issues/48464

Differential Revision: [D25292893](https://our.internmc.facebook.com/intern/diff/D25292893/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Use group.WORLD appropriately in process group initialization."


As part of investigating
https://github.com/pytorch/pytorch/issues/48464, I realized some weird
inconsistency in how we use `_default_pg` and `group.WORLD`. `group.WORLD`
apparently was an `object()` and never changed despite `_default_pg` changing.
In this sense, `group.WORLD` was being used a constant to refer to the default
pg, but wasn't of type PG at all. In fact the passed in group is also compared
via `==` to `group.WORLD` in many places, and it just worked since the default
argument was `group.WORLD`.

To clean this up, I got rid of `_default_pg` completely and instead used
`group.WORLD` as the default pg throughout the codebase. This also fixes the
documentation issues mentioned in
https://github.com/pytorch/pytorch/issues/48464.

#Closes: https://github.com/pytorch/pytorch/issues/48464

Differential Revision: [D25292893](https://our.internmc.facebook.com/intern/diff/D25292893/)

[ghstack-poisoned]
